lua/randomat2/events/mathracer.lua

```diff
@@ -0,0 +1,301 @@
+        end
+        tree["left"]["equation"] = tree["left"]["value"]
+        if tree["left"]["value"] < 0 then
+            tree["left"]["equation"] = "(" .. tree["left"]["equation"] .. ")"
+        end
+    end
+
+    if tree["right"]["value"] then
+        tree["right"] = self:SolveEquationTree(tree["right"]["value"])
+    else
+        if difficulty_options["termLowerBound"] < 0 then
+            if (math.random() < difficulty_options["negativeChance"]) then
+                tree["right"]["value"] = math.random(difficulty_options["termLowerBound"], -1)
+            else
+                tree["right"]["value"] = math.random(0 ,difficulty_options["termUpperBound"])
+            end
+        else
+            tree["right"]["value"] = math.random(difficulty_options["termLowerBound"], difficulty_options["termUpperBound"])
+        end
+        tree["right"]["equation"] = tree["right"]["value"]
+        if tree["right"]["value"] < 0 then
+            tree["right"]["equation"] = "(" .. tree["right"]["equation"] .. ")"
+        end
+    end
+
+    if tree["left"] == false or tree["right"] == false then return false end
+
+    local answer = operations[tree["operation"]]["calc"](tree["left"]["value"], tree["right"]["value"])
+    if not self:IsValidAnswer(answer) then return false end
+
+    local equation = tree["left"]["equation"] .. operations[tree["operation"]]["symbol"] .. tree["right"]["equation"]
+    if (tree["parentheses"]) then
+        equation = "(" .. equation .. ")"
+    end
+
+    return {
+        value = math.floor(answer),
+        equation = equation
+    }
+end
+
+function EVENT:GenerateEquation(first, time)
+    iteration = iteration + 1
+    if (iteration == 2) then
+        difficulty_options["terms"] = 3
+    elseif (iteration == 3) then
+        difficulty_options["termUpperBound"] = 12
+    elseif (iteration == 4) then
+        difficulty_options["termLowerBound"] = -12
+        difficulty_options["answerLowerBound"] = -100
+    elseif (iteration == 5) then
+        difficulty_options["terms"] = 4
+    elseif (iteration == 6) then
+        difficulty_options["answerLowerBound"] = -500
+        difficulty_options["answerUpperBound"] = 500
+    elseif (iteration == 7) then
+        difficulty_options["termLowerBound"] = -15
+        difficulty_options["termUpperBound"] = 15
+    elseif (iteration == 8) then
+        difficulty_options["terms"] = 5
+    elseif (iteration == 9) then
+        difficulty_options["negativeChance"] = 0.33
+        difficulty_options["answerLowerBound"] = -1000
+        difficulty_options["answerUpperBound"] = 1000
+    elseif (iteration >= 10 and iteration % 2 == 0) then
+        difficulty_options["terms"] = difficulty_options["terms"] + 1
+    end
+
+    local equation = self:SolveEquationTree(self:GenerateEquationTree(difficulty_options["terms"] - 1))
+    while equation == false do
```

Comment:
The result of `self:SolveEquationTree` is either a table or `false`. It can never be `nil` so we don't have to differentiate between `false`-y and `false`
```suggestion
    while not equation do
```